samza-core/src/main/java/org/apache/samza/storage/TransactionalStateTaskRestoreManager.java

```diff
@@ -0,0 +1,525 @@
+    Map<String, RestoreOffsets> storesToRestore = new HashMap<>();
+
+    storeEngines.forEach((storeName, storageEngine) -> {
+        // do nothing if store is non persistent and not logged (e.g. in memory cache only)
+        if (!storageEngine.getStoreProperties().isPersistedToDisk() &&
+          !storageEngine.getStoreProperties().isLoggedStore()) {
+          return;
+        }
+
+        // persistent but non-logged stores are always deleted
+        if (storageEngine.getStoreProperties().isPersistedToDisk() &&
+            !storageEngine.getStoreProperties().isLoggedStore()) {
+          File currentDir = storageManagerUtil.getTaskStoreDir(
+              nonLoggedStoreBaseDirectory, storeName, taskName, taskMode);
+          storeDirsToDelete.put(storeName, currentDir);
+          // persistent but non-logged stores should not have checkpoint dirs
+          return;
+        }
+
+        // get the oldest and newest current changelog SSP offsets as well as the checkpointed changelog SSP offset
+        SystemStream changelog = storeChangelogs.get(storeName);
+        SystemStreamPartition changelogSSP = new SystemStreamPartition(changelog, taskModel.getChangelogPartition());
+        SystemAdmin admin = systemAdmins.getSystemAdmin(changelogSSP.getSystem());
+        SystemStreamPartitionMetadata changelogSSPMetadata = currentChangelogOffsets.get(changelogSSP);
+        String oldestOffset = changelogSSPMetadata.getOldestOffset();
+        String newestOffset = changelogSSPMetadata.getNewestOffset();
+        String checkpointedOffset = checkpointedChangelogOffsets.get(changelogSSP);
+
+
+        Optional<File> currentDirOptional;
+        Optional<List<File>> checkpointDirsOptional;
+
+        if (!storageEngine.getStoreProperties().isPersistedToDisk()) {
+          currentDirOptional = Optional.empty();
+          checkpointDirsOptional = Optional.empty();
+        } else {
+          currentDirOptional = Optional.of(storageManagerUtil.getTaskStoreDir(
+              loggedStoreBaseDirectory, storeName, taskName, taskMode));
+          checkpointDirsOptional = Optional.of(storageManagerUtil.getTaskStoreCheckpointDirs(
+              loggedStoreBaseDirectory, storeName, taskName, taskMode));
+        }
+
+        LOG.info("For store: {} in task: {} got current dir: {}, checkpoint dirs: {}, checkpointed changelog offset: {}",
+            storeName, taskName, currentDirOptional, checkpointDirsOptional, checkpointedOffset);
+
+        // TODO BLOCKER pmaheshw: will do full restore from changelog even if retain existing state == true
+        // always delete current logged store dir for persistent stores.
+        currentDirOptional.ifPresent(currentDir -> storeDirsToDelete.put(storeName, currentDir));
+
+        // first check if checkpointed offset is invalid (i.e., out of range of current offsets, or null)
+        if (checkpointedOffset == null && oldestOffset != null) {
```

Comment:
Just FYI, current KafkaSystemAdmin does not follow the SSPMetadata interface, and returns oldest offset == 0 for an empty topic.

To your point, if you change the changelog topic for a store to an empty topic, this is the behavior:
1. Checkpointed offset will be null, since the checkpoint will not have an offset for the new changelog SSP.
2. Oldest offset may be non-zero, signifying that topic contains data, or null / 0, signifying that the topic is empty.
3. If oldest is non-zero, behavior depends on the retain.existing.changelog.state config. I renamed the config to clarify that if true, this will overwrite local state with the changelog topic state. This is also the case when you turn on the config for the first time, without writing changelog offests in checkpoints first.
4. If oldest is null or 0, it will fall through to the 'happy path', delete all local state, and do a full restore, which in this case will do nothing, which is the desired behavior.

I'll add an explicit unit test to cover the null oldest offset scenario.